### ISSUE TITLE
Make grid mixins respect media properties.

### ISSUE THREAD

**Issue by christianmagill** [thoughtbot/neat#591]
I feel that if the media property is part of a custom grid, then the grid-column, collapse, etc. mixins should respect the rule and only apply the related styles when the media condition is met.

**Comment by kgcreative** [comment]
This behavior is already true as long as you instance that object inside of a grid-media mixin.

**Comment by christianmagill** [comment]
Let's say I have a grid for featured items:

```
<div class='features'>
   <div class='feature'>...</div>
   <div class='feature'>...</div>
   <div class='feature'>...</div>
   <div class='feature'>...</div>
   <div class='feature'>...</div>
   <div class='feature'>...</div>
</div>
```

I want to setup a grid at 800px and that separates the features into 3 columns with 10px gutters.

```
$features-grid: (
   columns: 12,
   media: 800px,
   gutter: 10px,
);
```

I'd setup the items as such...

```
.features{
   @include grid-container();
}

.feature{
   @include grid-column($features);
   margin-bottom: 10px;
}

```
Currently regardless of whether the media condition is met, the feature class will receive the column settings. What I think is more appropriate is for them to only be applied when the condition specified is met.

So the code would in fact output:

```
.feature{
   @media (min-width: 800px){
      ... grid related rules ..
   }
   margin-bottom: 10px;
}
```

**Comment by christianmagill** [comment]
I understand that the behavior is true if you wrap all your mixins in a grid-media mixin, but if a mixin is taking a grid that already has the media rule, shouldn't it be respected?

**Comment by christianmagill** [comment]
Oh, sorry the grid-column parameter mixup was a typo on my part. 

I understand what you are saying and I like your grid-media implementation, I'm just wondering if there's any downside to making all the mixins respect the media property.

I'm already planning on making my own wrapper functions to extend functionality for my personal use, so I could implement there. It was just something that I expected to work so I thought I'd mention it.

**Comment by kgcreative** [comment]
I think it's definitely worth bringing up. Respecting the `media` property would actually greatly simplify markup. If the only reason you're emitting a `grid-media` is to instance a neat object (`grid-column`, `grid-container`, `grid-collapse`, etc) , then it would make sense to instance those with a media query and save yourself the grid media call.

@tysongach I could try figuring out a PR for this. My current thought would be a check for whether `$grid-media` is `$neat-grid`, and if so, retrieve `media` from the map and instance a media query

**Comment by christianmagill** [comment]
@kgcreative Thanks!  Couldn't you just look for the existence of "media" regardless of whether $grid-media is equal to $neat-grid.

The idea being that if a media property is specified it should be respected.

**Comment by kgcreative** [comment]
`grid-media` temporarily overrides `$neat-grid` with the custom grid and outputs @content inside a media query, so by checking for `$neat-grid` I can make the assumption a @media query is already being generated and we can avoid a double media entry. (Theoretically) otherwise I'd have to set a global variable when a neat grid is being generated, and that could get a little fragile

**Comment by kgcreative** [comment]
(Although maybe that's the solution) - I'll figure it out and report back :D

**Comment by wardpenney** [comment]
@kgcreative @christianmagill @tysongach Recently, we added an example about how to do a responsive grid to (the examples)[http://neat.bourbon.io/examples] and I think this thread is on the same page. 

But, `grid-media` it is really only a shortcut. It applies the block context of the grid passed-in to the code that runs in the block. Personally, I would just rather use the CSS `@media` directive because I don't mind passing my custom grids in the optional param.

**Comment by kgcreative** [comment]
@wardpenney - that's been my general though. I generally either pass my own @media directive, or use grid-media() as a shortcut. I'm somewhat hesitant to have a module output it's own media query if there is a media: element present, mostly because that seems to break the single-purpose aspect of most mixins, but at the same time, I see @christianmagill's point of having an expectation that if we're passing a $grid-media custom context that contains a media: setting, that the output should respect that.

**Comment by whmii** [comment]
I agree with @kgcreative that while this behavior would be super interesting, my biggest concern is that the majority of users would find this behavior surprising or confusing.

@christianmagill I think this would make a great extension that you could author. As mentioned in https://github.com/thoughtbot/neat/issues/592 we don't have a place for them yet but check out https://github.com/kgcreative/neat-omega as an example of what that might look like.

**Comment by christianmagill** [comment]
@whmii Perhaps, but I also think it could be confusing that the media property isn't respected, being that it's purpose is to define limits.

In order for someone to be confused with the alternative approach, they'd need to create a grid, decide to restrict via a media property, and then decide to use it outside of the limits they themselves have set.

I do like the idea of extensions, and will check out the example. Thanks!

**Comment by kgcreative** [comment]
@whmii - check out #596 - I'm not entirely sure this belongs in Core, although after testing it out, it is really nice. The wrapper function basically checks to see if you're inside a grid-media loop, and if so, then it doesn't emit a @media wrapper. Otherwise, if the passed grid contains a media: entry, it wraps that element in that media query.

I stopped myself from adding a setting that would conditionally wrap the element, but I could see adding an extra setting in either a grid media map or global neat settings that toggles the behavior.

This ended up being way less hacky than I thought.

**Comment by christianmagill** [comment]
@kgcreative Thanks for your work on this! I think a global setting is a good compromise.

**Comment by kgcreative** [comment]
@whmii I think this can be closed as `won't fix`. I haven't had the time to generalize this into an extension, and I'm personally not using this (or see the need for it, especially with CSS grid now out, which more or less means I now use neat as a fallback library more than anything else)
